ambry-replication/src/main/java/com.github.ambry.replication/CloudToStoreReplicationManager.java

```diff
@@ -0,0 +1,337 @@
+    public void onInstanceConfigChange(List<InstanceConfig> instanceConfigs, NotificationContext context) {
+      logger.debug("Instance config change notification received with instanceConfigs: {}", instanceConfigs);
+      ConcurrentSkipListSet<CloudDataNode> newVcrNodes = new ConcurrentSkipListSet<>();
+      ConcurrentHashMap<String, CloudDataNode> newInstanceNameToCloudDataNode = new ConcurrentHashMap<>();
+
+      // create a new list of available vcr nodes.
+      for (InstanceConfig instanceConfig : instanceConfigs) {
+        String instanceName = instanceConfig.getInstanceName();
+        Port sslPort =
+            getSslPortStr(instanceConfig) == null ? null : new Port(getSslPortStr(instanceConfig), PortType.SSL);
+        CloudDataNode cloudDataNode = new CloudDataNode(instanceConfig.getHostName(),
+            new Port(Integer.valueOf(instanceConfig.getPort()), PortType.PLAINTEXT), sslPort,
+            clusterMapConfig.clustermapVcrDatacenterName, clusterMapConfig);
+        newInstanceNameToCloudDataNode.put(instanceName, cloudDataNode);
+        newVcrNodes.add(cloudDataNode);
+      }
+
+      synchronized (notificationLock) {
+        instanceNameToCloudDataNode.set(newInstanceNameToCloudDataNode);
+        handleChangeInVcrNodes(newVcrNodes);
+      }
+    }
+  }
+
+  /**
+   * {@link LiveInstanceChangeListener} for vcr cluster.
+   */
+  private class LiveInstanceChangeListenerImpl implements LiveInstanceChangeListener {
+    @Override
+    public void onLiveInstanceChange(List<LiveInstance> liveInstances, NotificationContext changeContext) {
+      logger.debug("Live instance change notification received. liveInstances: {}", liveInstances);
+      ConcurrentSkipListSet<CloudDataNode> newVcrNodes = new ConcurrentSkipListSet<>();
+      // react to change in liveness of vcr nodes if the instance was earlier reported by helix as part of
+      // {@code onInstanceConfigChange} notification.
+      synchronized (notificationLock) {
+        for (LiveInstance liveInstance : liveInstances) {
+          if (instanceNameToCloudDataNode.get().containsKey(liveInstance.getInstanceName())) {
+            newVcrNodes.add(instanceNameToCloudDataNode.get().get(liveInstance.getInstanceName()));
+          }
+        }
+        handleChangeInVcrNodes(newVcrNodes);
+      }
+    }
+  }
+
+  /**
+   * {@link PartitionStateChangeListener} to capture changes in partition state.
+   */
+  private class PartitionStateChangeListenerImpl implements PartitionStateChangeListener {
+    @Override
+    public void onPartitionStateChangeToLeaderFromStandby(String partitionName) {
+      logger.debug("Partition state change notification from Standby to Leader received for partition {}",
+          partitionName);
+      synchronized (notificationLock) {
+        try {
+          addCloudReplica(partitionName);
+        } catch (ReplicationException rex) {
+          logger.error("Could not add replication for paritition {}", partitionName);
```

Comment:
NEED FEEDBACK: Note that here and for other listeners, I log an error in case the listener threw an exception, instead of throwing an exception here. With this approach, if anything fails, then cloud replication from only one partition or replica gets affected, and the notifications can keep coming for future change notifications. Let me know if you think this reasoning looks ok.

Reply:
Do you know in which case, it may throw exception?

Reply:
I think it's okay since it's a listener, but will the partition no longer be replicated from cloud?  Maybe we should add the partition to an error list and retry the failed ones later?  And/or update an error metric.

Reply:
ok, let's start with a metric here and implement possible re-try logic in future.